File: codewars/6kyu/valid-braces/dduck-valid-braces.py
# Valid Braces
# https://www.codewars.com/kata/5277c8a221e209d3f6000b56

from unittest import TestCase

import re

braces_detect_regexp = r'\(\)|\[\]|\{\}'


def validBraces(string):
    if re.search(braces_detect_regexp, string):
        return validBraces(re.sub(braces_detect_regexp, '', string))
    return len(string) == 0


TestCase().assertEqual(validBraces("(){}[]"), True)
TestCase().assertEqual(validBraces("([{}])"), True)
TestCase().assertEqual(validBraces("(}"), False)
TestCase().assertEqual(validBraces("[(])"), False)
TestCase().assertEqual(validBraces("[({})](]"), False)

File: codewars/6kyu/valid-braces/dduck-valid-braces.js
// Valid Braces
// https://www.codewars.com/kata/5277c8a221e209d3f6000b56

const bracesDetectRegExp = new RegExp(/\(\)|\[\]|\{\}/, 'g');

const validBraces = (braces = '') => {
  if (braces.match(bracesDetectRegExp)) {
    return validBraces(braces.replace(bracesDetectRegExp, ''));
  }
  return !braces.length;
};


console.log(validBraces("(){}[]"), true);
console.log(validBraces("([{}])"), true);
console.log(validBraces("(}"), false);
console.log(validBraces("[(])"), false);
console.log(validBraces("[({})](]"), false);
